cmd/share-createlink.go

```diff
@@ -0,0 +1,147 @@
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package cmd
+
+import (
+	"fmt"
+	"io/ioutil"
+	"os"
+	"os/user"
+	"path"
+	"path/filepath"
+	"strings"
+
+	"github.com/dropbox/dropbox-sdk-go-unofficial/dropbox/sharing"
+	"github.com/spf13/cobra"
+)
+
+/**
+** Try to get the share link for a file if it already exists.
+** If it doesn't make a new share link for it.
+ */
+func getShareLink(cmd *cobra.Command, args []string) (err error) {
+	if len(args) != 1 {
+		printShareLinkUsage()
+		return
+	}
+
+	dbx := sharing.New(config)
+	path, err := filepath.Abs(args[0])
+	if err != nil {
+		return
+	}
+
+	// Confirm that the file exists.
+	exists, err := exists(path)
+	if !exists || err != nil {
+		print("The file / folder specified does not exist.\n")
+		return
+	}
+
+	// Try to get a link if it already exists.
+	if getExistingLink(dbx, path) {
+		return
+	}
+
+	// print("File / folder does not yet have a sharelink, creating one...\n")
+
+	// The file had no share link, let's get it.
+	getNewLink(dbx, path)
+
+	return
+}
+
+func printShareLinkUsage() {
+	fmt.Printf("Usage: %s share getlink [file / folder path]\n", os.Args[0])
+}
+
+/*
+** Try to get an existing share link for a file / folder.
+** It returns true if the file / folder had a link. Otherwise it returns false.
+ */
+func getExistingLink(dbx sharing.Client, path string) bool {
+	// Remove the Dropbox folder from the start.
+	path = strings.Replace(path, getDropboxFolder(), "", 1)
+
+	arg := sharing.ListSharedLinksArg{Path: path}
+	// This method can be called with a path and just get that share link.
+	res, err := dbx.ListSharedLinks(&arg)
+	if err != nil || len(res.Links) == 0 {
+
+	} else {
+		printLinks(res.Links)
+		return true
+	}
+	return false
+}
+
+/*
+** Create and print a link for file / folder that doesn't yet have one.
+**
+** CreateSharedLinkWithSettings doesn't allow pending uploads,
+** so we use the partially deprecated CreateSharedLink.
+ */
+func getNewLink(dbx sharing.Client, path string) bool {
+	arg := sharing.NewCreateSharedLinkArg(strings.Replace(path, getDropboxFolder(), "", 1))
+	// Get the sharelink even if the file isn't fully uploaded yet.
+	arg.PendingUpload = new(sharing.PendingUploadMode)
+	// Determine whether the target is a file or folder.
+	fi, err := os.Stat(path)
+	if err != nil {
+		fmt.Println(err)
+		return false
+	}
+	switch mode := fi.Mode(); {
+	case mode.IsDir():
+		arg.PendingUpload.Tag = sharing.PendingUploadModeFolder
+	case mode.IsRegular():
+		arg.PendingUpload.Tag = sharing.PendingUploadModeFile
+	}
+	res, err := dbx.CreateSharedLink(arg)
+	if err != nil {
+		fmt.Printf("%+v\n", err)
+		return false
+	}
+	fmt.Printf("%s\t%s\n", res.Path[1:], res.Url)
+	return true
+}
+
+/* Return the path of the Dropbox folder. */
+func getDropboxFolder() string {
```

Comment:
This whole function is super hacky and likely not cross-platform. Let me think if there's a better way.

Reply:
Awesome, thanks. I'll address the other issues now and we can come back to this when you have a solution. There should be a function somewhere that returns where the Dropbox folder is, the information has to be held somewhere.

Reply:
Given the complexity around obtaining the "canonical" Dropbox folder, for all scenarios (e.g. paid users, paired accounts, custom directories) and for all platforms, I'd prefer if we solve this via documentation and help, and being explicit. For instance, we can ask users to just supply the path instead of trying to prefix things ourselves. WDYT?

Reply:
> pbcopy


Is this available in PR ?